components/uniset/src/utils.rs

```diff
@@ -0,0 +1,76 @@
+use std::{
+    char::MAX,
+    ops::{Bound::*, RangeBounds},
+};
+
+/// Returns whether the vector is sorted ascending non inclusive, of even length,
+/// and within the bounds of `0x0 -> 0x10FFFF` inclusive.
+pub fn is_valid(v: &[u32]) -> bool {
+    v.len() % 2 == 0
+        && v.chunks(2).all(|chunk| chunk[0] < chunk[1])
+        && v[v.len() - 1] <= (MAX as u32) + 1
+}
+
+/// Returns start (inclusive) and end (exclusive) bounds of RangeBounds
+pub fn deconstruct_range(range: &impl RangeBounds<char>) -> (u32, u32) {
+    let from = match range.start_bound() {
+        Included(b) => (*b as u32),
+        Excluded(b) => (*b as u32),
+        Unbounded => 0,
+    };
```

Comment:
Suggestion: you can fold into

```rust
let from = match range.start_bound() {
    Included(b) | Excluded(b) => (*b as u32),
    Unbounded => 0
};
```

Reply:
agreed and changed

Reply:
yeah, that sounds like a good improvement.

Reply:
Random note but as @zbraniecki mentioned with the new Rust 1.45.0 coming out today with changes to casting and procedural macros, the code in theory should now be able to be changed from:
```
let from = match range.start_bound() {
    Included(b) =>  (*b as u32),
    Excluded(b) => unreachable!(),
    Unbounded => 0
};
```
to
```
let from = match range.start_bound() {
    Included(b) =>  (*b as u32),
    unreachable!() => unreachable!(),
    Unbounded => 0
};
```
since the changes to procedural macros allow for them to be in pattern position now

Reply:
I think I prefer the former - it makes it easy to catch if `start_bound` would ever to return an enum with more variants.

Reply:
Ok, right now it is the former so I think we can resolve this as: agreed and changed